File: subsystems/android-device-connector/src/permission-manager.ts
/**
 * ==========================================================
 * SKOS
 * Smaily Knowledge Operating System
 * ==========================================================
 *
 * Subsystem : Android Device Connector
 * Module    : Permission Manager
 *
 * Build     : BUILD-000192
 * Version   : 1.0.0
 * ==========================================================
 */

/**
 * Supported Android permissions.
 */
export enum AndroidPermission {

    ReadInternalStorage = "read-internal-storage",

    ReadExternalStorage = "read-external-storage",

    ManageExternalStorage = "manage-external-storage"

}

/**
 * Permission state.
 */
export enum PermissionStatus {

    Granted = "granted",

    Denied = "denied",

    NotRequested = "not-requested"

}

/**
 * Permission result.
 */
export interface PermissionResult {

    permission: AndroidPermission;

    status: PermissionStatus;

}

/**
 * Permission manager.
 */
export class PermissionManager {

    /**
     * Check permission.
     *
     * Version 1.0
     * Placeholder implementation.
     */
    public check(
        permission: AndroidPermission
    ): PermissionResult {

        return {

            permission,

            status: PermissionStatus.NotRequested

        };

    }

}

File: SKOS-Android/app/src/main/java/com/smaily/skos/permission/PermissionManager.kt
package com.smaily.skos.permission

import android.content.Context
import android.content.pm.PackageManager
import androidx.core.content.ContextCompat

class PermissionManager(
    private val context: Context
) {

    fun checkPermission(permission: String): PermissionResult {

        val granted = ContextCompat.checkSelfPermission(
            context,
            permission
        ) == PackageManager.PERMISSION_GRANTED

        val state = if (granted) {
            PermissionState.GRANTED
        } else {
            PermissionState.DENIED
        }

        return PermissionResult(
            permission = permission,
            state = state,
            granted = granted
        )
    }

}
